### PR TITLE
[core] update cgroup v1 memory usage calculation to ignore inactive (cache) files

### PR DESCRIPTION
**Pull request by clarng**
Signed-off-by: Clarence Ng <clarence.wyng@gmail.com>

Adjust used memory calculation for cgroup v1, to make it inline with how working set memory is calculated, which is what the cgroup oom killer uses. Before this change we include the rss and cache, and not discount the inactive / purgeable cache content. When we write to disk or object store it generates a large amount of page cache. If we don't discount this cache content it will result in over-counting, and hence trigger the ray oom killer earlier than what it should be.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

Context
https://stackoverflow.com/questions/50865763/memory-usage-discrepancy-cgroup-memory-usage-in-bytes-vs-rss-inside-docker-con
https://utcc.utoronto.ca/~cks/space/blog/linux/OOMKillerWhen
https://engineering.linkedin.com/blog/2016/08/don_t-let-linux-control-groups-uncontrolled

### CODE REVIEW

**Comment by clarng** [comment]
Spin off this PR into smaller ones:

https://github.com/ray-project/ray/pull/29111/files
https://github.com/ray-project/ray/pull/29111/files

Those contain the changes to address comments here, so let's review those in favor of this - we might close this PR

@scv119  @stephanie-wang

File: src/ray/common/memory_monitor.cc [hunk 1]
```diff
@@ -95,6 +95,40 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
   return std::tuple(system_used_bytes, system_total_bytes);
 }
 
+int64_t MemoryMonitor::GetCGroupV1MemoryUsedBytes() {
```
Comment (scv119) [inline src/ray/common/memory_monitor.cc:98]: can we add a single unit test to test the parsing part? thanks!

**Review by scv119 — APPROVED** [review]

File: src/ray/common/memory_monitor.cc [hunk 2]
```diff
@@ -95,6 +95,40 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
...
+    } else if (title == "total_inactive_file") {
+      inactive_file_bytes = value;
+    }
+  }
+  if (rss_bytes == kNull || cache_bytes == kNull|| inactive_file_bytes == kNull) {
+    RAY_LOG_EVERY_MS(WARNING, kLogIntervalMs)
+      << "Failed to parse cgroup v1 mem stat. rss " << rss_bytes << " cache " << cache_bytes << " inactive " << inactive_file_bytes;
+      return kNull;
+  }
+  int64_t used = rss_bytes + cache_bytes - inactive_file_bytes;
```
Comment (scv119) [inline src/ray/common/memory_monitor.cc:128]: also might worth add reference to the links/docs/cadvisor code.

File: python/ray/tests/test_memory_pressure.py [hunk 3]
```diff
@@ -399,5 +402,40 @@ def test_newer_task_not_retriable_kill_older_retriable_task_first(
         ray.get(retriable_task_ref)
 
 
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_put_object_consume_shared_mem_and_page_cache_task_usage_slightly_below_limit_does_not_crash():
```
Comment (scv119) [inline python/ray/tests/test_memory_pressure.py:409]: @clarng looks linter is complaining this line too long.

**Pull request merged**